lib/spack/spack/build_systems/python.py

```diff
@@ -451,51 +475,52 @@ def build_directory(self):
         """
         return self.pkg.stage.source_path
 
-    def config_settings(self, spec, prefix):
+    def config_settings(self, spec: Spec, prefix: Prefix) -> Mapping[object, object]:
```

Comment:
Can we say something more than `object`? Also, why `object` and not `Any`?

Reply:
`object` and `Any` are opposites of each other. `Any` basically turns off the type checker. `object` means that all possible values must be supported. In this case, the only use of `config_settings` is that we convert the key/value pairs to strings, which all Python objects support. An example of this is something like:
```python
def config_settings(self, spec, prefix):
    return {
        "setup-args": {
            "-Dblas": "openblas",  # str
            "-Dparallel": make_jobs,  # int
            "-Ddebug": False,  # bool
        }
    }
```
There's no reason the keys/values have to be strings.

Reply:
See https://docs.python.org/3/library/typing.html#the-any-type

Reply:
Ok, I wasn't aware of the distinction, thanks. That said, I don't see why we can't be more specific.

> In this case, the only use of config_settings is that we convert the key/value pairs to strings, which all Python objects support

Sure, but consider the case in which users need to call a function themselves. If they know they deal with a string they can directly apply methods to values. Otherwise conversion is on them to do, before any manipulation.

Reply:
The user never needs to call these methods. In fact, at the moment, there isn't a single package that even overrides `config_settings` (soon numpy/scipy will). Of the 2,364 Python packages in Spack, exactly 0 call `self.import_modules`, `self.skip_modules`, `self.config_settings`, etc. These are only called by `PythonPackage`. We definitely want support for arbitrary objects here so that ints/floats/bools work as expected.

Reply:
You might be able to talk me into `Mapping[str, object]` but I definitely want support for ints/floats/bools.

Reply:
Union would be too difficult if we also want to support nested dicts. Although mypy does support recursive type definitions now, it's still too complicated for my taste.